posthog/api/csp.py

```diff
@@ -0,0 +1,233 @@
+import json
+from typing import Optional
+
+import structlog
+
+from posthog.exceptions import generate_exception_response
+from posthog.sampling import sample_on_property
+from posthog.utils_cors import cors_response
+from posthog.models.utils import uuid7
+
+from django.utils.html import escape
+
+logger = structlog.get_logger(__name__)
+
+"""
+| Normalized Key        | report-to format                     | report-uri format                  |
+| --------------------- | ------------------------------------ | ---------------------------------- |
+| `document_url`        | `body.documentURL`                   | `csp-report.document-uri`          |
+| `referrer`            | `body.referrer`                      | `csp-report.referrer`              |
+| `violated_directive`  | same as `effectiveDirective`         | `csp-report.violated-directive`    |
+| `effective_directive` | `body.effectiveDirective`            | `csp-report.effective-directive`   |
+| `original_policy`     | `body.originalPolicy`                | `csp-report.original-policy`       |
+| `disposition`         | `body.disposition`                   | `csp-report.disposition`           |
+| `blocked_url`         | `body.blockedURL`                    | `csp-report.blocked-uri`           |
+| `line_number`         | `body.lineNumber`                    | `csp-report.line-number`           |
+| `column_number`       | `body.columnNumber`                  | `csp-report.column-number`         |
+| `source_file`         | `body.sourceFile`                    | `csp-report.source-file`           |
+| `status_code`         | `body.statusCode`                    | `csp-report.status-code`           |
+| `script_sample`       | `body.sample`                        | `csp-report.script-sample`         |
+| `user_agent`          | top-level `user_agent`               | not available                      |
+| `report_type`         | top-level `type`                     | `"csp-violation"` constant         |
+"""
+
+
+def sample_csp_report(properties: dict, percent: float, add_metadata: bool = False) -> bool:
+    """
+    Sample CSP reports based on document_url and effective_directive
+
+    Args:
+        properties: The properties of the CSP report
+        percent: The sampling rate as a decimal (0.0 to 1.0)
+        add_metadata: Whether to add sampling metadata to the properties
+
+    Returns:
+        bool: True if the report should be included in the sampled set, False otherwise
+    """
+    if percent >= 1.0:
+        return True
+
+    document_url = properties.get("document_url", "")
+    effective_directive = properties.get("effective_directive", "")
```

Comment:
i'm tempted to say just document_url so i get all the reports for a page or not

but not-blocking we can easily tweak this later

Reply:
Oh, I thought we wanted to both last night, I'll change :)

Reply:
I agree! Never really understood the argument for the other one but you seemed convinced